src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,777 @@
+            let bb_idx = {
+                let bb_targets = &mut self.bb_targets;
+                let bb_target = &mut self.bb_target_count;
+                *bb_targets.entry((resume, drop)).or_insert_with(|| {
+                    let target = *bb_target;
+                    *bb_target = target.checked_add(1).unwrap();
+                    target
+                })
+            };
+            let source_info = data.terminator().source_info;
+            data.statements.push(self.set_state(bb_idx, source_info));
+            data.statements.push(Statement {
+                source_info,
+                kind: StatementKind::Assign(Lvalue::Local(RETURN_POINTER),
+                    self.make_state(state_idx, v)),
+            });
+            data.terminator.as_mut().unwrap().kind = TerminatorKind::Return;
+        });
+
+        self.super_basic_block_data(block, data);
+    }
+}
+
+fn get_body_id<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>, node_id: NodeId) -> (bool, hir::BodyId) {
+    // Figure out what primary body this item has.
+    match tcx.hir.get(node_id) {
+        hir::map::NodeItem(item) => {
+            match item.node {
+                hir::ItemConst(_, body) |
+                hir::ItemStatic(_, _, body) |
+                hir::ItemFn(.., body) => (false, body),
+                _ => bug!(),
+            }
+        }
+        hir::map::NodeTraitItem(item) => {
+            match item.node {
+                hir::TraitItemKind::Const(_, Some(body)) |
+                hir::TraitItemKind::Method(_,
+                    hir::TraitMethod::Provided(body)) => (false, body),
+                _ => bug!(),
+            }
+        }
+        hir::map::NodeImplItem(item) => {
+            match item.node {
+                hir::ImplItemKind::Const(_, body) |
+                hir::ImplItemKind::Method(_, body) => (false, body),
+                _ => bug!(),
+            }
+        }
+        hir::map::NodeExpr(expr) => {
+            // FIXME(eddyb) Closures should have separate
+            // function definition IDs and expression IDs.
+            // Type-checking should not let closures get
+            // this far in a constant position.
+            // Assume that everything other than closures
+            // is a constant "initializer" expression.
+            match expr.node {
+                hir::ExprClosure(_, _, body, _, _) => (true, body),
+                _ => (false, hir::BodyId { node_id: expr.id })
+            }
+        }
+        _ => bug!(),
+    }
+}
+
+fn ensure_generator_state_argument<'a, 'tcx>(
+                tcx: TyCtxt<'a, 'tcx, 'tcx>,
+                node_id: NodeId,
+                def_id: DefId,
+                mir: &mut Mir<'tcx>) -> (Ty<'tcx>, GeneratorInterior<'tcx>) {
+    let interior = *tcx.typeck_tables_of(def_id).generator_interiors.get(&node_id).unwrap();
+    let gen_ty = mir.local_decls.raw[1].ty;
+    (gen_ty, interior)
+}
+
+fn make_generator_state_argument_indirect<'a, 'tcx>(
```

Comment:
We allow moves out of the generator environment (since a generator only runs once), so it has to be  `Self` to allow that. This allows drop elaboration to partially drop the environment and MIR borrowck wouldn't complain about moves out of &mut.